File: ios/AppStarterKit/Core/Subscription/SubscriptionManager.swift
import Foundation

// RevenueCat integration stub
// TODO: Install RevenueCat SDK via Swift Package Manager:
//   .package(url: "https://github.com/RevenueCat/purchases-ios", from: "4.0.0")
//   then: import RevenueCat

enum SubscriptionTier: String {
    case free = "free"
    case tracker = "tracker"
}

enum TrackerFeature {
    case basicLogging, exerciseLibrary         // Always free
    case unlimitedTemplates, advancedAnalytics, exportData  // Tracker+
}

enum SubscriptionError: LocalizedError {
    case notConfigured

    var errorDescription: String? {
        "Subscription not configured. See SubscriptionManager.swift TODO comments."
    }
}

@Observable
@MainActor
final class SubscriptionManager {
    static let shared = SubscriptionManager()

    var currentTier: SubscriptionTier = .free
    var isLoading: Bool = false

    // TODO: Replace with your RevenueCat API key from the RevenueCat dashboard.
    private let apiKey = "YOUR_REVENUECAT_PUBLIC_API_KEY"

    /// Call once at app launch, after the user ID is known.
    /// - Parameter userId: The authenticated user's ID to associate purchases with.
    func configure(userId: String) {
        // TODO: Uncomment after adding RevenueCat SDK:
        // Purchases.logLevel = .debug
        // Purchases.configure(withAPIKey: apiKey, appUserID: userId)
    }

    /// Loads available offerings from RevenueCat.
    func fetchOfferings() async {
        isLoading = true
        // TODO: let offerings = try? await Purchases.shared.offerings()
        isLoading = false
    }

    /// Initiates a purchase for the given product ID.
    /// - Parameter productId: The RevenueCat product identifier (e.g. "tracker_monthly").
    func purchase(productId: String) async throws {
        // TODO: Resolve the matching Package from offerings, then:
        // let result = try await Purchases.shared.purchase(package: package)
        // if !result.userCancelled { currentTier = .tracker }
        throw SubscriptionError.notConfigured
    }

    /// Restores previous purchases for the current user.
    func restorePurchases() async throws {
        // TODO: try await Purchases.shared.restorePurchases()
    }

    /// Returns `true` when the user has access to the given feature.
    func isFeatureAvailable(_ feature: TrackerFeature) -> Bool {
        switch feature {
        case .basicLogging, .exerciseLibrary:
            return true  // Always available on free tier
        case .unlimitedTemplates, .advancedAnalytics, .exportData:
            return currentTier == .tracker
        }
    }
}

File: android/app/src/main/kotlin/com/appstarterkit/app/core/subscription/SubscriptionManager.kt
package com.appstarterkit.app.core.subscription

import android.content.Context
import dagger.hilt.android.qualifiers.ApplicationContext
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import javax.inject.Inject
import javax.inject.Singleton

// RevenueCat integration stub
// TODO: Add RevenueCat dependency to libs.versions.toml:
//   revenuecat = { group = "com.revenuecat.purchases", name = "purchases", version = "7.0.0" }
//   revenuecat-ui = { group = "com.revenuecat.purchases", name = "purchases-ui", version = "7.0.0" }
// Then add to build.gradle.kts: implementation(libs.revenuecat)

@Singleton
class SubscriptionManager @Inject constructor(
    @ApplicationContext private val context: Context,
) {
    // TODO: Replace with your RevenueCat public SDK key from dashboard
    private val sdkKey = "YOUR_REVENUECAT_PUBLIC_SDK_KEY"

    enum class SubscriptionTier { FREE, TRACKER }

    private val _currentTier = MutableStateFlow(SubscriptionTier.FREE)
    val currentTier: StateFlow<SubscriptionTier> = _currentTier.asStateFlow()

    fun configure(userId: String) {
        // TODO: Uncomment after adding SDK:
        // Purchases.logLevel = LogLevel.DEBUG
        // Purchases.configure(PurchasesConfiguration.Builder(context, sdkKey).appUserID(userId).build())
    }

    fun isFeatureAvailable(feature: TrackerFeature): Boolean = when (feature) {
        TrackerFeature.BASIC_LOGGING, TrackerFeature.EXERCISE_LIBRARY -> true
        TrackerFeature.UNLIMITED_TEMPLATES,
        TrackerFeature.ADVANCED_ANALYTICS,
        TrackerFeature.EXPORT_DATA,
        -> _currentTier.value == SubscriptionTier.TRACKER
    }

    enum class TrackerFeature {
        BASIC_LOGGING, EXERCISE_LIBRARY,           // Free
        UNLIMITED_TEMPLATES, ADVANCED_ANALYTICS, EXPORT_DATA, // Tracker+
    }
}
